File: src/test_swing/fmg/common/notyfier/NotifyPropertyChangedTest.java
package fmg.common.notyfier;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

import javax.swing.SwingUtilities;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import fmg.common.LoggerSimple;
import fmg.swing.utils.StaticInitializer;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

public class NotifyPropertyChangedTest {

    private INotifyPropertyChanged _dummy = new INotifyPropertyChanged() {

        @Override
        public void removeListener(PropertyChangeListener listener) { }

        @Override
        public void addListener(PropertyChangeListener listener) { }

    };

    @BeforeClass
    public static void setup() {
        StaticInitializer.init();
    }

    @Test
    public void NotifyPropertyChangedSyncTest() {
        try (NotifyPropertyChanged notifier = new NotifyPropertyChanged(_dummy, false)) {

            int countFiredEvents = 3 + ThreadLocalRandom.current().nextInt(10);
            int[] countReceivedEvents = { 0 };

            PropertyChangeListener listener = ev -> { ++countReceivedEvents[0]; };
            notifier.addListener(listener);
            for (int i=0; i<countFiredEvents; ++i)
                notifier.onPropertyChanged("propertyName");
            notifier.removeListener(listener);

            Assert.assertEquals(countFiredEvents, countReceivedEvents[0]);
        }
    }

    @Test
    public void NotifyPropertyChangedAsyncTest() throws InterruptedException {
        try (NotifyPropertyChanged notifier = new NotifyPropertyChanged(_dummy, true)) {

            int countFiredEvents = 3 + ThreadLocalRandom.current().nextInt(10);
            int[] countReceivedEvents = { 0 };
            Object[] firedValue = { null };

            Subject<PropertyChangeEvent> subject = PublishSubject.create();
            PropertyChangeListener listener = ev -> {
                ++countReceivedEvents[0];
                firedValue[0] = ev.getNewValue();
                subject.onNext(ev);
            };
            notifier.addListener(listener);

            final String prefix = " Value ";
            SwingUtilities.invokeLater(() -> {
                for (int i=0; i<countFiredEvents; ++i)
                    notifier.onPropertyChanged(null, prefix + i, "propertyName");
            });
            Signal signal = new Signal();
            Disposable dis = subject.timeout(50, TimeUnit.MILLISECONDS)
                    .subscribe(ev -> {
                        LoggerSimple.put("onNext: ev=" + ev);
                    }, ex -> {
                        LoggerSimple.put("onError: " + ex);
                        signal.set();
                    });


            Assert.assertTrue(signal.await(1000));

            notifier.removeListener(listener);
            dis.dispose();

            Assert.assertEquals(1, countReceivedEvents[0]);
            Assert.assertEquals(prefix + (countFiredEvents-1), firedValue[0]);
        }
    }

}

File: FastMines_core/UnitTest.FastMines.Core/Common/NotifyPropertyChangedTest.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Threading.Tasks;
using NUnit.Framework;
using fmg.common.ui;
using fmg.core.mosaic;

namespace fmg.common.notyfier {

    public class NotifyPropertyChangedTest {

        [OneTimeSetUp]
        public void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put(">" + nameof(NotifyPropertyChangedTest) + "::" + nameof(Setup));

            MosaicModelTest.StaticInitializer();

            //Observable.Just("UI factory inited...").Subscribe(LoggerSimple.Put);
        }

        [SetUp]
        public void Before() {
            LoggerSimple.Put("======================================================");
        }

        [OneTimeTearDown]
        public void After() {
            LoggerSimple.Put("======================================================");
            LoggerSimple.Put("< " + nameof(NotifyPropertyChangedTest) + " closed");
            LoggerSimple.Put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        [Test]
        public void NotifyPropertyChangedSyncTest() {
            var rand = new Random(Environment.TickCount);
            int countFiredEvents = 3 + rand.Next(10);
            int countReceivedEvents = 0;

            void listener(PropertyChangedEventArgs ev) { ++countReceivedEvents; }
            using (var notifier = new NotifyPropertyChanged(null, listener, false)) {
                for (int i = 0; i < countFiredEvents; ++i)
                    notifier.OnPropertyChanged("propertyName ");
            }
            Assert.AreEqual(countFiredEvents, countReceivedEvents);
        }

        [Test]
        public async Task NotifyPropertyChangedAsyncTest() {
            int countFiredEvents = 3 + new Random(Environment.TickCount).Next(10);
            int countReceivedEvents = 0;
            object firedValue = null;

            var subject = new Subject<PropertyChangedEventArgs>();
            void listener(PropertyChangedEventArgs ev) {
                ++countReceivedEvents;
                firedValue = (ev as PropertyChangedExEventArgs<string>).NewValue;
                subject.OnNext(ev);
            }
            var signalWait = false;
            const string prefix = "Value ";
            using (var notifier = new NotifyPropertyChanged(null, listener, true)) {
                using (var signal = new Signal()) {
                    using (subject.Timeout(TimeSpan.FromMilliseconds(100))
                        .Subscribe(ev => {
                            LoggerSimple.Put("OnNext: ev=" + ev);
                        }, ex => {
                            LoggerSimple.Put("OnError: " + ex);
                            signal.Set();
                        }))
                    {

                        for (var i=0; i<countFiredEvents; ++i)
                            notifier.OnPropertyChanged(null, prefix + i, "propertyName");

                        signalWait = await signal.Wait(TimeSpan.FromSeconds(1));
                    }
                }

            }

            LoggerSimple.Put("  " + nameof(NotifyPropertyChangedAsyncTest) + ": checking...");
            Assert.IsTrue(signalWait);
            Assert.AreEqual(1, countReceivedEvents);
            Assert.AreEqual(prefix + (countFiredEvents-1), firedValue);
        }

    }

}
